ctapipe/image/extractor.py

```diff
@@ -1285,3 +1288,158 @@ def __call__(
             peak_time=pulse_time2.astype("float32"),
             is_valid=is_valid,
         )
+
+
+@lru_cache
+def deconvolution_parameters(
+    camera: CameraDescription, upsampling: int, window_width: int, window_shift: int
+):
+    if upsampling < 1:
+        raise ValueError(f"upsampling must be > 0, got {upsampling}")
+    if window_width < 1:
+        raise ValueError(f"window_width must be > 0, got {window_width}")
+
+    ref_pulse_shapes = camera.readout.reference_pulse_shape
+    ref_sample_width = camera.readout.reference_pulse_sample_width.to_value("ns")
+    camera_sample_width = 1.0 / camera.readout.sampling_rate.to_value("GHz")
+
+    if camera_sample_width < ref_sample_width:
+        raise ValueError(
+            f"ref_sample_width (got {ref_sample_width}) must be equal to or shorter than camera_sample_width (got {camera_sample_width}); need a reference single p.e. pulse shape with finer sampling!"
+        )
+    avg_step = int(camera_sample_width / ref_sample_width + 0.5)
+
+    pzs = []
+    for ref_pulse_shape in ref_pulse_shapes:
+        phase_pzs = []
+        for phase in range(avg_step):
+            x = ref_pulse_shape[phase::avg_step]
+            i_min = np.argmin(np.diff(x)) + 1
+            if i_min < x.size and x[i_min] != 0:
+                phase_pzs.append(x[i_min + 1] / x[i_min])
+
+        if len(phase_pzs) == 0:
+            raise ValueError(
+                "ref_pulse_shape is malformed - cannot find deconvolution scale"
+            )
+        pzs.append(np.mean(phase_pzs))
+
+    gains, shifts = [], []
+    for pz, ref_pulse_shape in zip(pzs, ref_pulse_shapes):
+        integral = ref_pulse_shape.sum() * ref_sample_width / camera_sample_width
+        phase_gains, phase_shifts = [], []
+        for phase in range(avg_step):
+            x = ref_pulse_shape[phase::avg_step]
+            y = deconvolve(np.atleast_2d(x), 0.0, upsampling, pz)[0]
+            i_max = y.argmax()
+            start = i_max - window_shift  # TODO should use extract_around_peak here
+            stop = start + window_width
+            if start >= 0 and stop <= y.size:
+                phase_shifts.append(
+                    (i_max / upsampling * avg_step - ref_pulse_shape.argmax())
+                    * ref_sample_width
+                )
+                phase_gains.append(y[start:stop].sum() / integral)
+
+        if len(phase_gains) == 0:
+            raise ValueError(
+                "ref_pulse_shape is malformed - peak is not well contained"
+            )
+        gains.append(np.mean(phase_gains))
+        shifts.append(np.mean(phase_shifts))
+
+    return pzs, gains, shifts
+
+
+def deconvolve(waveforms, bls, up: int, pz: float):
```

Comment:
please add a docstring and avoid abbreviations (use clear variable names for  "up", "pz" and "bls")

Reply:
Got around to making a first attempt - is that useful enough?